mypy/checker.py

```diff
@@ -1705,7 +1733,7 @@ def check_return_stmt(self, s: ReturnStmt) -> None:
                         self.warn(messages.RETURN_ANY.format(return_type), s)
                     return
 
-                if self.is_unusable_type(return_type):
+                if self.is_unusable_type(return_type) or defn.is_async_generator:
```

Comment:
I think putting this in semanal.py would be hard because we only discover that the function is a generator during semantic analysis. I suppose we could do it in ThirdPass, but there are some code comments suggesting that ThirdPass may be removed in the future.